### ISSUE TITLE
Azure App Service Deploy

### ISSUE THREAD

**Issue by jukai16** [Microsoft/vsts-tasks#3112]
Hi,

I am trying to migrate my Azure App Service deployment from Classic (which is now marked as deprecated) to Azure App Service Deploy (RM). For some reason, I keep on getting:
*2016-11-18T00:39:16.1923786Z ##[error]Failed to deploy website.
2016-11-18T00:39:37.4268572Z ##[warning]Error: connect ETIMEDOUT [IP Removed]:443
2016-11-18T00:39:37.4268572Z ##[error]Error: Error: C:\Windows\system32\cmd.exe failed with return code: 4294967295*

A couple more details that may help:
- I am using an on-premise agent
- Task AzureRmWebAppDeploy version: 2.1.4
- may or may not matter but the webapp is hosted in a service plan using App Service Environment
Can someone help please?

Thanks in advance.

**Comment by jukai16** [comment]
An additional info,  there is an error.txt file created in the AzureRmWebAppDeploy task folder with the following content:

_Error Code: ERROR_DESTINATION_NOT_REACHABLE
More Information: Could not connect to the remote computer ("[webappname]scm.[asename].p.azurewebsites.net"). On the remote computer, make sure that Web Deploy is installed and that the required process ("Web Management Service") is started.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_DESTINATION_NOT_REACHABLE.
Error: Unable to connect to the remote server
Error: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond [AseIPAddress]:443
Error count: 1._

I have ports 454-455 open in the App Service Environment.

**Comment by rajatagrawal-dev** [comment]
Hi @jukai16 
Can you give more details as to how you are trying to migrate from Azure App Service : Classic to Azure App Service Deploy? Any app that you were previously able to deploy using the classic task should be able to deploy using the new task. Just make sure you are setting the correct subscription in the task.

**Comment by jukai16** [comment]
Hi @rajatagrawal-dev

Apologies, I should have replied to this earlier as soon as I was able to resolve it. The issue is not because of an incorrect subscription but rather the inbound security rules in the NSG.

In my case because my **App Service Environment** is created within a subnet with inbound rules to lockdown traffic, I have to explicitly open port 443. Luckily I am using on-premise build agent so I exactly know what the IP address is and only allow that IP as a source. Since opening this I was able to deploy successfully.

As far as hosted agents are concerned though (I have tested this as well just to make sure it wasn't my on-prem agent), using this task + ASE may not work well since the hosted agent's IP is unknown (please correct me if my assumption is incorrect). It kinda defeats the purpose of locking down inbound traffic if you have to open port 443 from _**"Any"**_ source should you decide to use a hosted agent to deploy to an App Service Environment.

**Comment by RoopeshNair** [comment]
Added it to our backlog. However, it will be 2-3 months before we can get to it based on current priorities.

Closing this for now.

Thanks,

**Comment by motowilliams** [comment]
Has the process of deploying to an App Service Environment from VSTS been resolved without having to deploy a jumpbox VM?

**Comment by akousmata** [comment]
I am seconding @motowilliams here.  Perhaps there's a way of making an NSG rule that allows the deployment utility being used to connect to the ASE?  I am using Azure DevOps Release Pipelines and everything works all the way up to the point where we need to deploy and can't because the server doesn't respond due to the fact that we have an ASE setup with an NSG that restricts inbound access.

**Comment by RoopeshNair** [comment]
> we have an ASE setup with an NSG that restricts inbound access

We have no way of reaching App Service that is behind an internal load balancer within an App Service Environment. The only option I can think of is using a jump-box or perhaps using Azure Relay.

**Comment by akousmata** [comment]
I found a way to allow this if you are ok with allowing all publicly facing Microsoft Azure Cloud IP's to have access.  In the NSG, set up a rule for a Source of Service Tag, and a Source service tag of AzureCloud or AzureCloud.<Region> with inbound access to ports 80 and 443. 

![source-service-tag](https://user-images.githubusercontent.com/5766227/52132310-7c618000-2604-11e9-8e08-b345311a7f46.PNG)

This worked for me and it may even work with some of the other Source service tags to get a more refined set of IP addresses.  It isn't perfect, but it's better than granting access to all of the internet and may not be secure enough for some people.  Just wanted to post this as an option.  Here is my final rule.

![image](https://user-images.githubusercontent.com/5766227/52132401-adda4b80-2604-11e9-805b-70a6717f1c33.png)
